### ISSUE TITLE
Overwriting mesh userData in GLTFLoader.

### ISSUE THREAD

**Issue by Tacohej** [mrdoob/three.js#15728]
##### Description of the problem

When a GLTF file is loaded with the GLTFLoader, mesh userData is overwritten by node extras if node extras exists.

* Function loadMesh gets called first meshDef.extras are assigned to mesh.userData at line 2757 in GLTFLoader.js:
`assignExtrasToUserData( mesh, meshDef );`

* Function loadNode gets called and mesh get assigned to mesh at line 3169:
`node = mesh;`

* Previously assigned userData gets overwritten at line 3218:
`assignExtrasToUserData( node, nodeDef );`

I don't know if I'm missing something, if this is expected behaviour or just a small bug. I expected it to merge node and mesh userData, or just use mesh userData.

##### Three.js version

- [x] Dev
- [x] r101
- [x] ... (It seems be be something that have been around for a while)

##### Browser

- [x] All of them
- [ ] Chrome
- [ ] Firefox
- [ ] Internet Explorer

##### OS

- [x] All of them
- [ ] Windows
- [ ] macOS
- [ ] Linux
- [ ] Android
- [ ] iOS

**Comment by donmccurdy** [comment]
> I expected it to merge node and mesh userData, or just use mesh userData.

If one has to take precedence, I definitely think node userData is more critical – meshes may be reused across many nodes, and each node could have different user data.

Merging sounds OK to me, although note that it won't happen in every case – nodes containing multiple primitives will turn into a `THREE.Group`, not a `THREE.Mesh`, in some cases.

**Comment by Tacohej** [comment]
> Merging sounds OK to me, although note that it won't happen in every case – nodes containing multiple primitives will turn into a `THREE.Group`, not a `THREE.Mesh`, in some cases.

That's true. In that case, every mesh in a group would get it's own userData from primitives, and the parent node would get its userData right?

I don't know if merging is a better approach, but at least no data would be lost.

**Comment by mrdoob** [comment]
How about:

```js
object.userData.gltfNodeExtra = {};
object.userData.gltfMeshExtra = {};
```

**Comment by donmccurdy** [comment]
> every mesh in a group would get it's own userData from primitives, and the parent node would get its userData right?

Well, sometimes. If the mesh only has one primitive, we flatten that into a single Mesh, rather than creating a `Group > Mesh` pair. Or sometimes primitives can be merged into a single BufferGeometry using groups. In general it's hard to have the threejs scene graph match exactly the internal glTF representation. We try to stay close, but also to be efficient.

Maybe this is as good a time as any to revisit merging primitives into a single geometry with groups... removing that simplifies things a little. @takahirox do you have a preference on it? Is it still important for other cases, do you think?

I like the simplicity and intuitiveness of `Object.assign( object.userData, mesh.extras, node.extras )`. But it does leave the possibility of overwriting keys with the same names, and I'm a bit torn on whether preventing collisions is worth the verbosity of `gltf<Type>Extras` namespacing.

**Comment by donmccurdy** [comment]
> Maybe this is as good a time as any to revisit merging primitives into a single geometry with groups...

^This has been removed. I think we can just merge the user data now, something equivalent to;

```js
Object.assign( object.userData, mesh.extras, node.extras )
```

**Comment by donmccurdy** [comment]
This is done:

https://github.com/mrdoob/three.js/blob/master/examples/jsm/loaders/GLTFLoader.js#L2016
